feat(maitake): add `WaitQueue`

hawkw opened the pull request:
this branch adds a `WaitQueue` type in `maitake::wait`. this allows
multiple tasks to wait for an event, and be woken either one at a time
(in FIFO fashion), or all at once.

there are additional APIs we could add here, but this is a good starting
place. there probably should be docs examples but i'm lazy.

jamesmunns commented on maitake/src/wait/queue.rs:
```diff
@@ -0,0 +1,790 @@
+            // if there are no waiters in the queue, increment the number of
+            // `wake_all` calls and return.
+            State::Woken | State::Empty => {
+                self.state.fetch_add(QueueState::ONE_WAKE_ALL, SeqCst);
```
Could this be made AcqRel?

hawkw replied:
tried changing all atomic ops to AcqRel/Acquire, and one of the loom tests broke; i think there are at least some cases where SeqCst is necessary (perhaps to synchronize with the mutex?). will investigate further in a follow up.

jamesmunns commented on maitake/src/wait/queue.rs:
```diff
@@ -0,0 +1,790 @@
+
+        // okay, we actually have to wake some stuff.
+
+        // TODO(eliza): wake outside the lock using an array, a la
```
Can this todo be closed? I don't really track...

hawkw replied:
this is a thing i still want to do in a follow-up PR.

jamesmunns commented on maitake/src/wait/queue.rs:
```diff
@@ -0,0 +1,790 @@
+    /// synchronization primitives or resources: when an event makes a resource
+    /// permanently unavailable, the queue can be closed.
+    pub fn close(&self) {
+        let state = self.state.fetch_or(State::Closed as u8 as usize, SeqCst);
```
Might wanna put a big ol' warning flag on the State variable that this is dependent on "Closed" being "all bits set" of the relevant state fields

jamesmunns commented on maitake/src/wait/queue.rs:
```diff
@@ -0,0 +1,790 @@
+    ///
+    /// # Safety
+    ///
+    /// This is only safe to call while the list is locked. The dummy `_list`
```
The comment say it's a dummy, but `list` is used inside the function

jamesmunns commented on maitake/src/wait/queue/tests.rs:
```diff
@@ -0,0 +1,298 @@
+    use core::sync::atomic::{AtomicUsize, Ordering};
+
+    #[test]
+    fn wake_all() {
```
This would be awesome to steal as an example at the top level docs.

jamesmunns reviewed (commented):
Added some high level comments - I admit to not totally grokking how everything works yet, but I'm going to go try and integrate this branch into mnemos now, so I might have more meaningful feedback after that.

hawkw commented on maitake/src/wait/queue.rs:
```diff
@@ -0,0 +1,790 @@
+    /// The node's waker
+    waker: Wakeup,
+
+    // /// Optional user data.
```
this is commented out because i punted on implementing user data in this branch, i should probably just delete it.

hawkw commented on maitake/src/wait/queue.rs:
```diff
@@ -0,0 +1,790 @@
+#[derive(Debug)]
+#[repr(C)]
+#[pin_project]
+struct Waiter {
```
i don't really think this is a `Header`; in the case of the task header, the naming is because the header is a small amount of data stored before the task's "body" (the future/output). in this case, the `Waiter` type is the entirety of the data in the linked list...

hawkw commented on maitake/src/wait/queue.rs:
```diff
@@ -0,0 +1,790 @@
+
+/// The state of a [`Waiter`] node in a [`WaitQueue`].
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+enum WaitState {
```
30/60 bits is definitely enough, because it always comes from the value stored in `QueueState`, which is also missing 2 bits...

hawkw replied:
done in 8cc540eab56064e313b9c5b533869f436c764ef1

hawkw commented:
@jamesmunns i think i've addressed most of your review feedback; if you're still working on integrating this in mnemOS, i'd be happy to wait for that, but i  have a slight preference for merging this now and addressing any additional issues that come up in a follow-up, if that's okay?

jamesmunns commented:
Misc note (can totally be a future PR): It might be nice to be able to modify the lock behavior somehow.

For me, any lock contention is likely to be a programming bug, so `.try_lock().unwrap()` is likely to make a "louder"/easier to catch case for me.

jamesmunns commented:
> @jamesmunns i think i've addressed most of your review feedback; if you're still working on integrating this in mnemOS, i'd be happy to wait for that, but i have a slight preference for merging this now and addressing any additional issues that come up in a follow-up, if that's okay?

It's integrated and working, no complaints from use! I'll take a quick scan of the recent changes and r+ it.

hawkw commented:
> 



> Misc note (can totally be a future PR): It might be nice to be able to modify the lock behavior somehow.
> 
> For me, any lock contention is likely to be a programming bug, so `.try_lock().unwrap()` is likely to make a "louder"/easier to catch case for me.

hmm, this isn't a bad idea, but it might be kind of painful to figure out a good interface for overriding lock behavior...could be worth opening a separate issue for?

jamesmunns commented on maitake/src/wait/queue.rs:
```diff
@@ -0,0 +1,927 @@
+
+    /// Wake the next task in the queue.
+    ///
+    /// If the queue is empty, a wakeup is stored in the `WaitQueue`, and the
```
Two things I don't grok:

What happens if I call "wake" a bunch with an empty queue?

How come this stores a "wakeup" in the queue, rather than keeping it at the "empty" state?

hawkw replied:
it's necessary to store a single wakeup in the queue because a task may be attempting to enqueue itself *while* `wake` is called, which would result in a "lost" wake. multiple calls to `wake` on an empty queue will not store additional wakeups; storing a count of wakeups in the queue seems like a job for a semaphore, which i'd also like to add to `maitake`.

jamesmunns reviewed (approved):
Overall LGTM! Added one "I don't understand" question, but that's mostly on me, unless you want to update the docs to explain.
